### DISCUSSION TITLE
Bulk Update selected rows with value

### DISCUSSION THREAD

**Discussion by benjamin-steurer** [Budibase/budibase#11055] [Help] [answered]
I am trying to select multiple rows, press a button which then prompts me for a value. After I have entered the value, this value should be saved back to a specific column in the table.

So in this example I would like to change the "Estimated Sales" to 30. So I would want to select both rows, the Button should prompt me for a value and after I entered 30 and clicked save, the value of "Estimated Sales" should change to 30
<img width="1054" alt="image" src="https://github.com/Budibase/budibase/assets/117825209/1a77a9b7-8a71-4674-b4f8-6ac630ab5107">

This is a similar open Issue:
https://github.com/Budibase/budibase/discussions/5666

**Comment by ConorWebb96** [comment]
Hey @Blochergroup,

I believe in this instance you could use automations for this along with the automation app action.

I've created a simple demo below to help guide you on how to do this.

1. Go to your automations area
2. Add a new automation, name it whatever you want. (In my case I called it Bulk Update)
<img width="100%" alt="Screenshot 2023-06-29 at 10 56 26" src="https://github.com/Budibase/budibase/assets/126772285/2f0edab1-43cc-4791-8492-1f934e80a8b0">
3. Select App action as the automation type and click the save button.
4. Add in 2 different fields which will be used later. One being the rows for the select rows array that you will pass to the automation. The other is up to you. In my case I went with the name newFieldValue.
<img width="100%" alt="Screenshot 2023-06-29 at 10 30 14" src="https://github.com/Budibase/budibase/assets/126772285/f9926c05-cd2b-4532-88d9-9d18486b0c1c">
5. Add a new automation step and select the `Update Row` option and then click save.
<img width="100%" alt="Screenshot 2023-06-29 at 10 30 26" src="https://github.com/Budibase/budibase/assets/126772285/91812328-0389-4d4f-bddb-1b86a1b35b88">
6. Click the add looping button.
<img width="100%" alt="Screenshot 2023-06-29 at 10 30 54" src="https://github.com/Budibase/budibase/assets/126772285/144aa3cf-ab4b-4f43-aa4d-b0c135c7272d">
7. Select string for the input type
<img width="100%" alt="Screenshot 2023-06-29 at 10 31 09" src="https://github.com/Budibase/budibase/assets/126772285/dc2aaa00-c66f-4e0c-89b8-9d1f90e2bc59">
8. Use the field created previously here for the binding/value
<img width="100%" alt="Screenshot 2023-06-29 at 10 31 28" src="https://github.com/Budibase/budibase/assets/126772285/29d4df9e-3f79-43ad-849c-5f74f46bef09">
9. Select the table you wish to run this automation on.
<img width="100%" alt="Screenshot 2023-06-29 at 10 31 43" src="https://github.com/Budibase/budibase/assets/126772285/e7d607ff-3fc5-48d3-907a-d3712e5fe9a3">
10. Target the value you wish to change and set it to the `newFieldValue` created earlier on in the fields area. Additionally, add `loop.currentItem` to the row id so the automation knows which row to update.
<img width="100%" alt="Screenshot 2023-06-29 at 10 32 27" src="https://github.com/Budibase/budibase/assets/126772285/cd28aaeb-c0b8-4bd0-abc5-44f28e9c05ba">

11. That finishes up everything on the automation, next head over to the design section.
12. Add a button, then add the run automation action to this when its clicked.
13. Add a input field used to send the value you want the fields to change to.
<img width="100%" alt="Screenshot 2023-06-29 at 10 33 59" src="https://github.com/Budibase/budibase/assets/126772285/3a017743-333d-4d80-9d00-a6ba5cf8fa40">
14. Configure the run automation action similar to the below.
<img width="100%" alt="Screenshot 2023-06-29 at 10 34 13" src="https://github.com/Budibase/budibase/assets/126772285/651375fa-b35d-4796-88ed-825c58c5ce22">
15. Publish the app and that should be it working.

**Demo Below**
![Loom Message - 29 June 2023](https://github.com/Budibase/budibase/assets/126772285/89e61a5e-50b1-4053-b8db-401ffa5d7bfd)

I hope this helps!

**Reply by benjamin-steurer** [reply]
Hey @ConorWebb96 ,
first of all thank you for your help!

Unfortunately I can't get this to work. Here are my current settings:
My Automation:
<img width="196" alt="image" src="https://github.com/Budibase/budibase/assets/117825209/5f830e7f-7edd-480f-97a6-7076f31fcbf0">

Custom Value (Input Field):
<img width="199" alt="image" src="https://github.com/Budibase/budibase/assets/117825209/c2b52e94-2e6c-4db1-a8f0-20fa5df7796c">

My Button:
<img width="730" alt="image" src="https://github.com/Budibase/budibase/assets/117825209/fe7c910e-d697-44c8-96c3-80c35a55ee70">

I am a bit unsure about the CustomValue (Input Field) but the rest should be pretty right.
Thank You!

**Reply by ConorWebb96** [reply]
Hey @Blochergroup,

Sorry to hear that, have you definitely published your app? 

Are you getting any errors whenever click the button to run the automation? Can you click the 3 bars on the far right of the screen and let me know if the automation has run/has errors?

For the input, its not needed you can just type in the value inside without linking to the field, publish then view the production app, select your rows and click the submit button. Make sure any changes you make are published, as automations only work in published apps.

**Reply by benjamin-steurer** [reply]
Hey @ConorWebb96 ,
the app is definitely published.
I am getting "Automation triggered" after clicking the Button but the values just don't change.
I also just looked at the Automation History and it just seems like the value of the Field isn't passed to the automation.

Could you maybe send a screenshot of the settings you have within your Text Input Field?
I feel like I am doing something wrong with it because the automation of course has to "know" which field is "newFieldValue"

**Reply by ConorWebb96** [reply]
My bad I didn't realise I forgot to supply the screenshot for the input!
<img width="313" alt="Screenshot 2023-06-30 at 08 55 35" src="https://github.com/Budibase/budibase/assets/126772285/3503357b-4d8a-4d57-86e5-74d321701384">

In the below screenshot, you should be able to see the binding for the input when attaching it in the automation action area.
<img width="318" alt="Screenshot 2023-06-30 at 08 55 54" src="https://github.com/Budibase/budibase/assets/126772285/1115b599-a946-4c79-9d6e-76ea44807d7c">

I hope this helps, let me know if this helps sort the issue.

**Reply by benjamin-steurer** [reply]
Thank you for the Help @ConorWebb96 !
It seems like I had mixed up a text input with a number field.
I fixed that and I can see the Value I did input in that field in the Automation History. So far so good!

I still have one error with the "Update Row" Action
<img width="237" alt="image" src="https://github.com/Budibase/budibase/assets/117825209/607b4001-10ee-49af-a8e9-0baebf63fb97">

"UID can't be blank".
UID is the primary Key of my MS SQL table but I don't understand the error as we use the row ID to update the row and the UID shouldn't  be necessary, right?

**Reply by sklaffa** [reply]
I also have the very same issue. I'm trying to do the very same thing, that is, bulk update a field in multiple rows. It looks like the "update row" action always tries to update all the fields of the row, even if there is no need to change their values. So, if some of these fields is actually a foreign key to a row of another table, the update fails because that field is required. In my opinion, the expected behaviour of "update row" should be something like "just update the fields for which you were given a value, and let the others alone".

**Reply by ConorWebb96** [reply]
Hey @Blochergroup,

Apologies I figured out the issue you were having and forgot to post the answer to this for you. Basically your MsSQL scehma has its own validation on this field. Therefore, we need to pass the existing values back in the automation. This does mean you will have to change the automation slightly from what I demoed previously. The changes required are below; 

1. Add a `query rows` step before the `update row` step
2. Select the table for the rows you wish to query.
<img width="100%" alt="Screenshot 2023-07-04 at 07 30 45" src="https://github.com/Budibase/budibase/assets/126772285/c1fa704d-d2c4-4abf-87b4-4a878b31a1e1">
3. Click the define filters button, and customise it similar to the below
<img width="100%" alt="Screenshot 2023-07-04 at 07 30 52" src="https://github.com/Budibase/budibase/assets/126772285/f059ea01-b21c-4848-9d9c-bd54bafcdb3f">
4. Updated your looping type from string to array and passing in the `{{ step.1.rows }}` from the previous step
<img width="100%" alt="Screenshot 2023-07-04 at 07 31 28" src="https://github.com/Budibase/budibase/assets/126772285/1ea6dab1-08fb-4895-b1b3-5bf34dd836d7">
5. Then filling the the values of the required fields using the syntax `{{ loop.currentItem.nameOfField }}`
<img width="100%" alt="Screenshot 2023-07-04 at 07 32 34" src="https://github.com/Budibase/budibase/assets/126772285/6801dd05-7506-4b71-908e-ba585eebc69b">
6. Publish the changes and run the automation


I hope this helps resolve your issue!

**Reply by sklaffa** [reply]
Hello, for me it still does not work. Here are the screenshots:

STEP 1: configure the automation
![step1](https://github.com/Budibase/budibase/assets/104690977/e48f6cd1-b923-4e96-be54-4d16fc0b91f2)

STEP 2: configure the filter of the "query rows" action
![step2](https://github.com/Budibase/budibase/assets/104690977/f121300e-8aaf-47ba-b5ed-6f384f504790)

STEP 3: configure the looping
![step3](https://github.com/Budibase/budibase/assets/104690977/6cbe5702-a1bc-4318-a27f-72cb1c7c5e83)

STEP 4: configure the "update row" action
![step4](https://github.com/Budibase/budibase/assets/104690977/4d799db0-a440-47d7-9c2e-0d64035d9721)

RESULT: checking the "automation history" it can be seen that the list of selected rows contains "rogue" characters (they seem to be the UFT-8 codes for the square brackets), thus forcing the query to fail and return no rows for the subsequent loop:
![step5](https://github.com/Budibase/budibase/assets/104690977/fc01b026-fdf3-405f-8a78-0a62a15feae7)

EDIT: I noticed that in my case the "selected_rows" trigger field was set as "text" type instead of "array". I tried to switch to "array" and tested again, but looking at the automation history the result is exactly the same, and the string of selected rows still contains the unexpected characters.

How can I solve this?

**Comment by andz-bb** [comment, marked as answer]
hey @Blochergroup @sklaffa 

I've come up with a new solution that should work well for you both. It's similar to Conor's method above but instead of querying the rows in the automation, I am passing an array of full row objects to the automation from the data provider. I am using a Postgres table to test this so it should work with most other external datasources.

In order to access that full row data I have made sure that the update button where I'm triggering the automation is nested within the data provider as shown here:
<img width="258" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/953fa744-4ec1-4d53-a991-19da08e166e4">
  

Here are the actions I have for the button:
<img width="885" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/b44fa61b-72f0-486d-9785-d65e9539154c">

Here is the custom JS I am using to get the value for rows:
```js
return $("Dataprovider.Rows").filter(x => $("Table.Selected rows").includes(x._id));
```
This returns an array of the full row objects for the ids that are selected in the table.

Now that we are passing the full objects to the array we can reduce the complexity of the automation. Here are the updated automation steps:

For the trigger values I am making sure that `rows` is set to array type
<img width="521" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/b4fc1a31-89c7-4a80-ab16-da4e2e85161b">

Then I have an Update Row step which loops over the `rows` array on the trigger. I am using `id` rather than `_id` for my Row ID since this is a Postgres datasource.
<img width="600" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/914f3f8a-bdc8-4b13-b127-215ad2d51b38">

Finally I have a [Collect Data step](https://docs.budibase.com/docs/synchronous-automations) at the end of my automation. This is not necessary for the basic functionality of this but is a nice bonus as it allows my automation to run synchronously. This means that the Refresh Data Provider action on the update button on my screen will only run after my automation has finished updating the table.
<img width="498" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/0758cadd-5fb2-432c-9ebd-9d63e19f0eec">

Here's the app in action. I hope this helps!
![bulk update](https://github.com/Budibase/budibase/assets/110921612/ec47de93-222f-4d4a-90ba-99007716c2ef)

**Reply by benjamin-steurer** [reply]
@andz-bb thank you for your solution!
This is a interesting way to solve the problem. Your solution is working for me up until the actual SQL update. 
In the Update Row step of the automation:
```
"success": false,
      "response": "Error: RequestError: Violation of PRIMARY KEY constraint 'PK__SKUSales__C5B196023CE24CF2'. Cannot insert duplicate key in object 'dbo.SKUSalesData'. The duplicate key value is *UID*
```

It seems like the Update Row Action is trying to create a new row instead of updating the existing rows?
Would be great to get the input of @sklaffa because he is also using Postgres. Maybe its just an error with MS SQL?

**Reply by andz-bb** [reply]
@Blochergroup could you send a copy of your DB schema and an app export to support@budibase.com please? It would be really useful to have a look on my end. There's a [new feature in the latest version of Budibase (v2.8) that generates a schema for you](https://github.com/Budibase/budibase/discussions/11055#discussioncomment-6370952) which will hopefully make the process easier. Thanks

**Reply by benjamin-steurer** [reply]
@andz-bb just got it to work :) 
Seems like I had a typo in the _id field step in the automations. 

Thank you so much, that will save a lot of clicks & time in the future.

**Comment by sklaffa** [comment]
*** EDIT ***
Deleting the automation and reimplementing it from scratch did the trick. I don't know why the previous one didn't work, but the new one does. Thanks for the contributions and above all for the patience 👍 

************


Hi All, sorry to reopen this thread after so long, but I didn't have the opportunity to test this solution earlier due to a variety of reasons (both professional and personal). So, long story short, **it still does not work** for me. But this time the problem is different in my case. Implementing the solution suggested by @andz-bb I managed to correctly pass the selected rows to the automation, but for some reason I can't understand, the loop block in the automation receives an empty input and exits with "no iterations".

This is how I modified my automation, as per @andz-bb instructions:

![budibase_bug_2](https://github.com/Budibase/budibase/assets/104690977/da34e51c-6f48-4cfc-a673-ad9bbbc9d293)

Here is the log of the automation execution as seen from the "manage" section of my app, as you can see the output of the trigger block correctly contains the array of the selected rows, but the input to the looping block is empty so no rows are actually updated:

![budibase_bug_1](https://github.com/Budibase/budibase/assets/104690977/4e9546f5-765a-43ed-9867-de1ec0630834)

I'm stuck at this point and I am unsure what to do next. Maybe I should delete and re-create the automation? The app is updated to the latest version (2.9.25), and the backend database is MariaDB (but considering the automation logs I don't think that the backend is relevant for the problem I'm experiencing).

**Comment by Jednadvacet** [comment]
Hello, I followed the automation update guide above and I succeeded, but I have problems with relationship typed columns.

My automation trigger updates boolean fields "Approved" from false to true, as intended. But updates relationship field "Workers" to empty as well, although the checkbox "Clear relationships if empty?" is not checked.

Seems to me like a bug. Can anyone work it around?

(PostgresDB, latest docker image)

**Reply by ConorWebb96** [reply]
Hey @Jednadvacet, 

Have you tried re-passing the relationship values in that are already attached to the record?

If you could, would it be possible to write up a [issue](https://github.com/Budibase/budibase/issues/new/choose). I thought I had written this bug up but wasn't able to find it.

If you can't I'll try and get an example etc written up later today. Thanks for reminding me!

Thanks ahead of time!

**Comment by Hades1280** [comment]
Hi, iam trying to resolve this, but without positive results.

Automation seems to work rightly.

![Captura de pantalla 2025-02-17 152109](https://github.com/user-attachments/assets/865a971e-5d4c-4e1e-9865-36144c2d94f4)
![Captura de pantalla 2025-02-17 152129](https://github.com/user-attachments/assets/e88744d2-cba6-4ad3-9908-3bd62943b5ef)
![Captura de pantalla 2025-02-17 153335](https://github.com/user-attachments/assets/f7d60af7-e04e-4218-b0ac-11056fcfd816)
![Captura de pantalla 2025-02-17 153348](https://github.com/user-attachments/assets/feba4984-654e-42c9-9604-c4c80b941d8a)

Components are correct, only difference is i try to use a option picker.

![Captura de pantalla 2025-02-17 152150](https://github.com/user-attachments/assets/84b5a558-fadb-4c8a-9264-9ea082392dcb)
![Captura de pantalla 2025-02-17 152219](https://github.com/user-attachments/assets/7dfad8fd-9f7e-4ac7-9540-15a1fada8864)
![Captura de pantalla 2025-02-17 152448](https://github.com/user-attachments/assets/8b958cb7-9ba0-4046-ac43-4f2e4e126942)
![Captura de pantalla 2025-02-17 152438](https://github.com/user-attachments/assets/962b478d-5de4-4017-89fb-b4f0df0eb355)
![Captura de pantalla 2025-02-17 152406](https://github.com/user-attachments/assets/ccd0eaf0-2581-4f18-9fc8-9e76f585d8b0)

Button settings aparently fine.

![Captura de pantalla 2025-02-17 152248](https://github.com/user-attachments/assets/c93b9b7b-97a2-4ec0-9df9-7ed2009ac5d0)

Using js like @andz-bb suggest.

![Captura de pantalla 2025-02-17 152303](https://github.com/user-attachments/assets/f764ac73-51f6-488f-9c47-6c14b267f87c)



![Captura de pantalla 2025-02-17 152327](https://github.com/user-attachments/assets/2c286018-6893-4476-af32-2bbc3a143489)


Chrome show this when ia try to update "Estado".
![Captura de pantalla 2025-02-17 153648](https://github.com/user-attachments/assets/2a20cb68-c266-4e3f-b509-a931e7698f3a)

Iam using docker image version 3.4.7.

Any help will be appreciated.
